tensorboard/webapp/persistent_settings/_data_source/persistent_settings_data_source_test.ts

```diff
@@ -0,0 +1,233 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {Injectable} from '@angular/core';
+import {TestBed} from '@angular/core/testing';
+import {firstValueFrom} from 'rxjs';
+
+import {
+  LocalStorageTestingModule,
+  TestingLocalStorage,
+} from '../../util/local_storage_testing';
+import {
+  OSSSettingsConverter,
+  PersistentSettingsDataSourceImpl,
+  SettingsCoverter,
+  TEST_ONLY,
+} from './persistent_settings_data_source';
+import {BackendSettings, PersistableSettings} from './types';
+
+describe('persistent_settings data_source test', () => {
+  let localStorage: TestingLocalStorage;
+  let getItemSpy: jasmine.Spy;
+
+  async function configure<UiSettings, StorageSettings>(
+    Converter: new () => SettingsCoverter<UiSettings, StorageSettings>
+  ) {
+    await TestBed.configureTestingModule({
+      imports: [LocalStorageTestingModule],
+      providers: [
+        PersistentSettingsDataSourceImpl,
+        {provide: SettingsCoverter, useClass: Converter},
```

Comment:
why provider here only "PersistentSettingsDataSourceImpl" but not
```
   {
      provide: PersistentSettingsDataSource,
      useClass: PersistentSettingsDataSourceImpl,
    },
```

Reply:
Great question. Below, we `TestBed.inject`  the `PersistentSettingsDataSourceImpl` and only test the implementation of `PersistentSettingsDataSourceImpl`. Whether we provide it under `PersistentSettingsDataSource` InjectionToken or not does not matter for this test.